src/ui/src/builder/useComponentActions.ts

```diff
@@ -1134,6 +1134,91 @@ export function useComponentActions(
 		return page.id;
 	}
 
+	/**
+	 * Extracts components from a blueprint for publishing.
+	 * Excludes the blueprint container itself, notes, and internal components.
+	 */
+	function extractBlueprintComponents(blueprintId: Component["id"]): Component[] {
+		const subtree = getFlatComponentSubtree(blueprintId);
+		return subtree
+			.slice(1) // exclude the blueprint container itself
+			.filter((c) => c.type !== "note")
+			.filter((c) => {
+				const definition = wf.getComponentDefinition(c.type);
+				return definition?.category !== "Internal";
+			});
```

Comment:
It's better to use a single iteration
```suggestion
			.filter((c) => c.type !== "note" && wf.getComponentDefinition(c.type)?.category !== "Internal");
```